pkg/controller/keycloakclient/keycloakclient_controller.go

```diff
@@ -0,0 +1,306 @@
+		authenticated, err := r.getAuthenticatedClient(keycloak)
+		if err != nil {
+			return r.ManageError(instance, err)
+		}
+
+		// Compute the current state of the realm
+		log.Info(fmt.Sprintf("got authenticated client for keycloak at %v", keycloak.Status.InternalURL))
+		clientState := common.NewClientState(r.context, keycloak)
+
+		log.Info(fmt.Sprintf("read client state for keycloak %v/%v, realm %v/%v, client %v/%v",
+			keycloak.Namespace,
+			keycloak.Name,
+			instance.Namespace,
+			instance.Spec.RealmSelector.MatchExact,
+			keycloak.Namespace,
+			instance.Spec.Client.ClientID))
+
+		err = clientState.Read(r.context, instance, authenticated, r.client)
+		if err != nil {
+			return r.ManageError(instance, err)
+		}
+
+		// Figure out the actions to keep the realms up to date with
+		// the desired state
+		reconciler := NewKeycloakClientReconciler(keycloak)
+		desiredState := reconciler.Reconcile(clientState, instance)
+		actionRunner := common.NewRealmActionRunner(r.context, r.client, r.scheme, instance, authenticated)
+
+		// Run all actions to keep the realms updated
+		err = actionRunner.RunAll(desiredState)
+		if err != nil {
+			return r.ManageError(instance, err)
+		}
+	}
+
+	return reconcile.Result{Requeue: false}, r.manageSuccess(instance, instance.DeletionTimestamp != nil)
+}
+
+// Fills the CR with default values. Nils are not acceptable for Kubernetes.
+func (r *ReconcileKeycloakClient) adjustCR(cr *kc.KeycloakClient) {
+	if cr.Spec.Client.ID == "" {
+		cr.Spec.Client.ID = cr.Namespace + "-" + cr.Name
+		log.Info(fmt.Sprintf("client ID is not specified, using %v", cr.Spec.Client.ID))
+	}
+	if cr.Spec.Client.Attributes == nil {
+		cr.Spec.Client.Attributes = make(map[string]string)
+	}
+	if cr.Spec.Client.Access == nil {
+		cr.Spec.Client.Access = make(map[string]bool)
+	}
+}
+
+func (r *ReconcileKeycloakClient) validate(instance *kc.KeycloakClient) error {
+	// If no selector is set we can't figure out which Keycloak instance this realm should
+	// be added to. Skip reconcile until a selector has been set.
+	if instance.Spec.InstanceSelector == nil {
+		return fmt.Errorf("realm %v/%v has no instance selector and will be ignored", instance.Namespace, instance.Name)
+	}
+	// If no realm selector is set we can't figure out which Keycloak Client we need to target for the API calls
+	if instance.Spec.RealmSelector == nil || instance.Spec.RealmSelector.MatchExact == "" {
+		return fmt.Errorf("client %v/%v has no realm selector and will be ignored", instance.Namespace, instance.Name)
+	}
+	// If there's no client id, we can't process this any further
+	if instance.Spec.Client == nil {
+		return fmt.Errorf("client %v/%v has no Client definition and will be ignored", instance.Namespace, instance.Name)
+	}
+	if instance.Spec.Client.ClientID == "" {
+		return fmt.Errorf("client %v/%v has no Client ID and will be ignored", instance.Namespace, instance.Name)
+	}
+	return nil
+}
```

Comment:
This can all be done using kubebuilder markers in the types file. The benefit of this is that the user should get an immediate response that they have specified something wrong in the Spec rather than success on the `create` and an async failure through the status.

see https://book.kubebuilder.io/reference/markers/crd-validation.html

Reply:
Good catch! Fixed.